### PR TITLE
Rework Dimensions for 1.16

### PR DESCRIPTION
**Pull request by modmuss50**
Its possible for mods to add dimensions without any code, and without any input from fabric using vanilla's new datapack system. This PR currenly re-adds the custom entity teleportation stuff that was present in the old API, in an almost exact manner.

The example mod does add a dimension using vanilla's new system, however it marks the world as experimental. I tried to remove mod added dims from this check, but got lost in DFU. My idea was an api such as `FabricDimensions.registerStableDimension`. 

I think the best course of action is to move this PR into a custom module for the custom portal placement logic as this can be used on its own, or with vanilla datapack added dimensions.

### CODE REVIEW

**Comment by Daomephsta** [comment]
What's with the loader and mapping update?

**Comment by modmuss50** [comment]
Mappings were updated as they provided some mappings for the new dim stuff.

**Review by jaskarth — APPROVED** [review]

**Comment by modmuss50** [comment]
Ready for review, I have reduced the scope of this PR to just replace what was lost.

File: fabric-dimensions-v1/src/testmod/java/net/fabricmc/fabric/test/dimension/FabricDimensionTest.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		try {
+			ServerPlayerEntity serverPlayerEntity = context.getSource().getPlayer();
+			ServerWorld serverWorld = serverPlayerEntity.getServerWorld();
+
+			if (!serverWorld.method_27983().equals(dimensionRegistryKey)) {
+				FabricDimensions.teleport(serverPlayerEntity, dimensionRegistryKey, FabricDimensionTest::placeEntity);
+			} else {
+				FabricDimensions.teleport(serverPlayerEntity, DimensionType.OVERWORLD_REGISTRY_KEY, FabricDimensionTest::placeEntity);
+			}
+		} catch (CommandSyntaxException e) {
```
Comment (i509VCB) [inline fabric-dimensions-v1/src/testmod/java/net/fabricmc/fabric/test/dimension/FabricDimensionTest.java]: Don't catch the command syntax exception. Let brigadier handle it (you can make the method throw it and be fine).

Reason being the command manager will return a richer message when you attempt to get the ServerPlayerEntity if you are running as a server and fail for you.

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FabricDimensionInternals.java [hunk 2]
```diff
@@ -39,12 +46,14 @@ private FabricDimensionInternals() {
 	public static final boolean DEBUG = System.getProperty("fabric.dimension.debug", "false").equalsIgnoreCase("true");
 	public static final Logger LOGGER = LogManager.getLogger();
 
+	public static final LinkedHashMap<RegistryKey<DimensionType>, Pair<DimensionType, ChunkGeneratorFactory>> FABRIC_DIM_MAP = new LinkedHashMap<>();
```
Comment (i509VCB) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/FabricDimensionInternals.java]:
```suggestion
	public static final Map<RegistryKey<DimensionType>, Pair<DimensionType, ChunkGeneratorFactory>> FABRIC_DIM_MAP = new LinkedHashMap<>();
```

**Comment by modmuss50** [comment]
Updated again, please provide reviews.

I have tested this an it allows creating dimensions for vanilla clients to join.

**Comment by kvverti** [comment]
Looks like you need to put license headers in

**Comment by i509VCB** [comment]
Ah yes the old `licenseFormat`

**Comment by modmuss50** [comment]
Got it adding dims in pre 1 now, still some major issues such as it not adding them to existing worlds, and showing the experimental world screen with modded dimensions.

**Comment by modmuss50** [comment]
Its possible for mods to add dimensions without any code, and without any input from fabric using vanilla's new datapack system. This PR currenly re-adds the custom entity teleportation stuff that was present in the old API, in an almost exact manner.

The example mod does add a dimension using vanilla's new system, however it marks the world as experimental. I tried to remove mod added dims from this check, but got lost in DFU. My idea was an api such as `FabricDimensions.registerStableDimension`. 



I think the best course of action is to move this PR into a custom module for the custom portal placement logic as this can be used on its own, or with vanilla datapack added dimensions.

File: fabric-dimensions-v1/src/testmod/java/net/fabricmc/fabric/test/dimension/FabricDimensionTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	private int executeTestCommand(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
+		ServerPlayerEntity serverPlayerEntity = context.getSource().getPlayer();
+		ServerWorld serverWorld = serverPlayerEntity.getServerWorld();
+
+		if (!serverWorld.getRegistryKey().equals(dimensionRegistryKey)) {
+			serverPlayerEntity.changeDimension(serverWorld.getServer().getWorld(dimensionRegistryKey));
+		} else {
+			FabricDimensions.teleport(serverPlayerEntity, serverWorld.getServer().getWorld(World.OVERWORLD), FabricDimensionTest::placeEntity);
```
Comment (i509VCB) [inline fabric-dimensions-v1/src/testmod/java/net/fabricmc/fabric/test/dimension/FabricDimensionTest.java]: can be inlined to `context.getSource().getServer()`

File: settings.gradle [hunk 4]
```diff
@@ -44,5 +44,5 @@ include 'fabric-rendering-fluids-v1'
 include 'fabric-resource-loader-v0'
 include 'fabric-tag-extensions-v0'
 include 'fabric-textures-v0'
-//include 'fabric-dimensions-v1'
+include 'fabric-dimensions-v1'
```
Comment (i509VCB) [inline settings.gradle]: I think we should move this up to be in alphabetical order like how all the other modules appear

File: fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension/fabric_dimension/void.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (i509VCB) [inline fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension/fabric_dimension/void.json:1]: I don't think these files should be in the minecraft namespace?

Shouldn't this be located at `src/testmod/resources/data/testrmod/dimension/fabric_dimension/void.json`

Comment (Daomephsta) [reply fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension/fabric_dimension/void.json:1]: It's a quirk of the system.
> Unlike other data pack contents, dimensions only function in the `minecraft` namespace, and require a sub folder which is treated as the namespace. For example `data/minecraft/dimension/custom/test.json` will be called with `custom:test`.

https://minecraft.gamepedia.com/Dimension#Dimension_syntax

Comment (i509VCB) [reply fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension/fabric_dimension/void.json:1]: ah okay so this is a special case

Comment (modmuss50) [reply fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension/fabric_dimension/void.json:1]: Yes.. this confused me at first, almost seems like a vanilla bug IMO.

**Review by Player3324 — APPROVED** [review]

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java [hunk 6]
```diff
@@ -23,16 +23,18 @@
...
+ * used in Entity::changeDimension.
+ *
+ * @deprecated Experimental feature, may be removed or changed without further notice due to potential changes to Dimensions in subsequent versions.
  *
  * @see FabricDimensions
- * @see FabricDimensionType
  */
+@Deprecated
 @FunctionalInterface
 public interface EntityPlacer {
```
Comment (reoseah) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: how about something like `DimensionTeleportHandler` or just `TeleportHandler`? `EntityPlacer` is very vague and it leads to vague class names in user code

File: fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (LemmaEOF) [inline fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json:1]: Example for fabric entity placer? also, does this need to go in a `minecraft/` subfolder like the dimension JSON?

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java (hunk 6 above)
Comment (reoseah) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: Right before this place in javadoc word `used` is not capitalized at the start of a sentence

File: fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json (hunk 7 above)
Comment (modmuss50) [reply fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json:1]: The entity placer is handled in code, see the example mod.

And yes this also needs to go in `minecraft/`

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java (hunk 6 above)
Comment (modmuss50) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: This doesnt actually have anything to do with the teleportation. Its just handling the spawn (location, and possibly portal creation) of the entity in the other world.

File: fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json (hunk 7 above)
Comment (LemmaEOF) [reply fabric-dimensions-v1/src/testmod/resources/data/minecraft/dimension_type/fabric_dimension/void_type.json:1]: ahhh, alright, sounds good

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java (hunk 6 above)
Comment (reoseah) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: `TeleportListener` or `TeleportPlacer`?

Comment (reoseah) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: For me, `EntityPlacer` can be anything related to entity spawning, entities summoning other entities for attack etc
`TeleportationSomething` is significantly clearer. 
Anything that mentions "teleport" or "teleportation" will be better than that

Comment (modmuss50) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/api/dimension/v1/EntityPlacer.java:34]: `TeleportListener` is wrong, possibly `TeleportPlacer`?

**Pull request merged**